adagrams/game.py

```diff
@@ -1,11 +1,121 @@
+#from tests.test_wave_01 import LETTER_POOL
+import random
+import copy
+LETTER_POOL = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+
+SCORE_DICT = {
+    1: ['A', 'E', 'I', 'O', 'U', 'L', 'N', 'R', 'S', 'T'],
+    2: ['D', 'G'],
+    3: ['B', 'C', 'M', 'P'],
+    4: ['F', 'H', 'V', 'W', 'Y'],
+    5: ['K'],
+    8: ['J', 'X'],
+    10: ['Q', 'Z'],
+}
+
 def draw_letters():
-    pass
+    # generate list containing all letters
+    # in corresponding frequency
+    all_letters = []
+    for letter,frequency in LETTER_POOL.items():
+        for i in range(frequency):
+            all_letters.append(letter)
+    
+    hand = []
+    for i in range(10):
+        this_letter = random.choice(all_letters)
+        hand.append(this_letter)
+        all_letters.remove(this_letter)
+
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    word_bank = copy.deepcopy(letter_bank)
+    caps_word = word.upper()
+    for letter in caps_word:
+        if letter not in word_bank:
+            return False    
+        else: 
+            word_bank.remove(letter)    
+    else:
+        return True
+
+
 
 def score_word(word):
-    pass
+    total = 0
+
+    for letter in word.upper():
+        for value, letter_list in SCORE_DICT.items():
+            if letter in letter_list:
+                total += value
```

Comment:
Here's where there's a bit of a mismatch between the data we are trying to look up (a letter) and the key in our structure. This approach essentially requires that we (in the worst case) iterate over the entire letter score structure to find each letter. Now given that our code always expects the SCORE_DICT to be constant, we could write off the whole thing as a constant term (26 letters to check max, and therefore this whole thing is something like O(26n) → O(n)) and not worry about it. But it's also worth taking a step back.

Consider what would happen to our analysis if this `score_word` were instead a helper function `score_word_helper(word, letter_scores)` and our `score_word` looked more like
```python
def score_word(word):
    return score_word_helper(word, SCORE_DICT)
```

Now our `score_word_helper` has two parameters that could vary in length, so now we have a nested loop with two sets of data, and our efficiency is revealed to be O(n*m) or again, something more like O(n^2) in the worst case.

So if we are committed to the idea that our function will only ever have to work with the scores in SCORE_DICT, we could get away with calling this O(n), but if we step back and recognize that those scores are just another piece of data, we have to admit that this looks more like O(n^2).

Think about how we could improve this by rearranging how we store SCORE_DICT.